backend/httpclient/provider.go

```diff
@@ -0,0 +1,204 @@
+package httpclient
+
+import (
+	"crypto/tls"
+	"net/http"
+)
+
+// Provider provides abilities to create http.Client, http.RoundTripper and tls.Config.
+type Provider interface {
```

Comment:
Looking at this interface I have a couple of questions:
1) Is its main responsibility is only to provide a `*http.Client` in the end? In this perspective - do we really need three different methods here instead of one that returns a ready to use HTTP client? Sorry if I missed some obvious  idea here
2) Is this interface really required? Since having only a concrete type seems enough. Consumers can still mock behaviour on consumer side if required.

Reply:
1. No, there's use cases for retrieving a http.Client, http.RoundTripper or tls.Config. The idea is that this will replace the following support in Grafana: https://github.com/grafana/grafana/blob/643e7af3e0395859803f956f442a3e0ed6dc68db/pkg/models/datasource_cache.go#L138-L259
In addition, core HTTP backend data sources in Grafana can be converted to use the SDK with this functionality.
2. The interface is mainly needed in Grafana or at least right now. The idea is to configure a provider in Grafana at startup and to reuse this throughout the whole life-cycle of application until shutdown, providing http.Client functionality for anyone needing it. See https://github.com/grafana/grafana/pull/33439. Same could apply for an external backend plugin that would like to apply the same middleware/settings throughout the lifetime without having to provide them as options everytime creating a new http.Client, http.RoundTripper or tls.Config.

Reply:
Got it 👍 - found some places in Grafana core where only RoundTripper or tls.Config used.

The bad thing in non-generic interfaces like this is that we will need to think twice before extending or changing it in the future. Since we don't need different implementations here (or we need?) it's simpler to deal with a struct. This way we can always add new methods in the future since we have not committed on interface. See also:

* https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_let_callers_define_the_interface_they_require - this gives a good tip about defining required interfaces on consumer side to clearly say what subset of methods expected by a consumer. For example if MySQL client only requires TLS config it can define TLSProvider interface on caller side.
* https://www.ardanlabs.com/blog/2016/10/avoid-interface-pollution.html - this gives a tip that interface with the only one really useful implementation can be unnecessary

Since the amount of code and number of use cases for this client client is big enough I can be wrong for this particular use case - just hope you will consider this points. Also, I am not really sure how this fits current approaches inside Grafana core since it has some established practices with interface usage and dependency resolution.

Reply:
Yeah that's actually something I've been thinking of. Might move the Provider interface to Grafana instead for now. Thanks for valuable feedback

Reply:
Done